### PR TITLE
Refined and Implemented Control Library

### PR DESCRIPTION
**Pull request by collinsc**
# Prequsite
#41 
# Description
We added the control library a while ago, but we didn't have a great understanding of it's inner workings. We also didn't have any testing on real world sensors. 

This pull request I took our arduino architecture and integrated it with the control system. I also hooked it up with an encoder and the PID controllers and identified a repeatable procedure for tuning our PID when the time comes using rqt.

The main chances came from relabling the components and reorganizing them to match the system described in this roscon talk, by one of the authors of the control library. 
https://vimeo.com/107507546

I relabled `controller.cpp` to `robot interface`, because  it techinically resides in the hardware layer, and is a misnomer for it's commanding layer the control layer.

I renamed `controller launcher` to `controller.cpp`, because it manages the control layer, (there is an external entity that handles launching controllers)

I also fixed some narrowing conversions in drivebase. The parameter server uses doubles, and it uses floats, and control want's doubles. 

# TODO
Next we need to implement emergency stop, but this pull request was getting big, and I wanted to break it up.

Also we need to get some potentiometers hooked up. These two items will be on a separate pull request.

# Testing 
Obviously you can't run this. So video demo.
https://youtu.be/z_qjUdycNio

### CODE REVIEW

File: src/tfr_control/include/tfr_control/pwm_interface.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+{
+    #include <gpio.h>
+}
+
+class PWMInterface
+{
+    public:
+        enum class Address
+        {
+            LEFT_TREAD = 0
```
Comment (RyanBerge) [inline src/tfr_control/include/tfr_control/pwm_interface.h]: Either add the other components to this list or add a TODO comment for that task.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
Just a small request.

File: src/tfr_control/arduino/encoder/encoder.ino [hunk 2]
```diff
@@ -1,46 +1,41 @@
...
 }
 
 void loop()
 {
-    velocities.timestamp = nh.now();
-    velocities.left_vel = gearbox_left.getVelocity()/GEARBOX_RPM;
-    drivebase_publisher.publish(&velocities);
-    nh.spinOnce();
+    reading.left_vel = gearbox_left.getVelocity()/GEARBOX_RPM;
+    //TODO hook up right encoder
```
Comment (AdamuKaapan) [inline src/tfr_control/arduino/encoder/encoder.ino]: Also todo, hook up turntable (just making sure this is known, you don't need to worry about adding a comment or anything)

File: src/tfr_control/config/controllers.yaml [hunk 3]
```diff
@@ -1,51 +1,60 @@
...
+# The controller manager handles the upper level logic, and makes sure
+# the controllers do not conflict
 # ------------------------------------------------------------
 
 # Publish all joint states -----------------------------------
 joint_state_controller:
-  type: joint_state_controller/JointStateController
-  publish_rate: 50
+    type: joint_state_controller/JointStateController
+publish_rate: 50
```
Comment (AdamuKaapan) [inline src/tfr_control/config/controllers.yaml]: Indentation check, this got unindented: is this correct, does that change any functionality?

**Review by RyanBerge — DISMISSED** [review]

File: src/tfr_control/include/tfr_control/robot_interface.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        // Populated by us for controller layer to use
+        double position_values[JOINT_COUNT]{};
+        // Populated by us for controller layer to use
+        double velocity_values[JOINT_COUNT]{};
+        // Populated by us for controller layer to use
+        double effort_values[JOINT_COUNT]{};
+
+        /**
+         * Adam uses this for testing the arm
+         * TODO delete
```
Comment (AdamuKaapan) [inline src/tfr_control/include/tfr_control/robot_interface.h]: Talked with you about this in person, I'll work with you on changing it to match the better style later, I'd prefer for it to stay in for now.

File: src/tfr_control/launch/control.launch [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <!-- Load all of the motor controllers -->
   <rosparam file="$(find tfr_control)/config/controllers.yaml" command="load"/>
 
-  <!-- Load the URDF into the ROS Parameter Server, processing macros as we
+  <!-- TODO adam delete Load the URDF into the ROS Parameter Server, processing macros as we
```
Comment (AdamuKaapan) [inline src/tfr_control/launch/control.launch]: This is irrelevant here, this line is actually needed.

**Review by AdamuKaapan — CHANGES_REQUESTED** [review]
Just had one (relevant) comment about checking indentation in a file, making sure it doesn't break anything. The other comments are just about little notes.

**Review by AdamuKaapan — APPROVED** [review]

**Pull request merged**